pkg/virt-launcher/virtwrap/sriov/hostdev_test.go

```diff
@@ -30,231 +35,394 @@ import (
-		}
-		Expect(devices, err).To(Equal([]api.HostDevice{expectHostDevice1}))
-	})
+		It("fails due to a missing event from a sriov device", func() {
+			var done bool
+			initErr := fmt.Errorf("initialized error")
+			err := initErr
+
+			hostDevice2 := api.HostDevice{Alias: &api.Alias{Name: sriov.AliasPrefix + "device2"}}
+			r := &registrarStub{failDeregister: true}
+			f := func() error { done = true; return nil }
+			go func() {
+				err = sriov.WaitHostDevicesToDetach(
+					r, []api.HostDevice{hostDevice, hostDevice2}, 2*time.Second, f)
+			}()
+
+			Eventually(func() interface{} {
+				return r.callback
+			}).ShouldNot(BeNil(), "callback has not been registered")
+
+			callback := r.callback.(libvirt.DomainEventDeviceRemovedCallback)
+			callback(nil, nil, &libvirt.DomainEventDeviceRemoved{DevAlias: api.UserAliasPrefix + "non-sriov"})
+			callback(nil, nil, &libvirt.DomainEventDeviceRemoved{DevAlias: api.UserAliasPrefix + sriov.AliasPrefix})
+
+			Expect(done).To(BeTrue())
+			Eventually(func() error { return err }, 3*time.Second).ShouldNot(SatisfyAll(MatchError(initErr), BeNil()))
+		})
 
-	It("creates 1 device that includes boot-order", func() {
-		net := newNetworkData("net1", "resource1", "0000:81:01.0", "0000:81:01:1")
+		// Failure to deregister the callback only emits a logging error.
+		It("succeeds to wait for a detached device and fails to deregister a callback", func() {
+			var done bool
+			err := fmt.Errorf("initialized error")
 
-		iface := newSRIOVInterface(net.Name)
-		val := uint(1)
-		iface.BootOrder = &val
-		vmi := &v1.VirtualMachineInstance{}
-		vmi.Spec.Domain.Devices.Interfaces = []v1.Interface{iface}
+			r := &registrarStub{failDeregister: true}
+			f := func() error { done = true; return nil }
+			go func() { err = sriov.WaitHostDevicesToDetach(r, []api.HostDevice{hostDevice}, 2*time.Second, f) }()
 
-		var devices []api.HostDevice
-		var err error
-		env := []envData{net.ResourceEnv, net.DeviceEnv}
-		withEnvironmentContext(env, func() {
-			devices, err = sriov.CreateHostDevices(vmi)
-		})
+			Eventually(func() interface{} {
+				return r.callback
+			}).ShouldNot(BeNil(), "callback has not been registered")
 
-		hostPCIAddress1 := api.Address{Type: "pci", Domain: "0x0000", Bus: "0x81", Slot: "0x01", Function: "0x0"}
-		expectHostDevice1 := api.HostDevice{
-			Alias:     &api.Alias{Name: sriov.AliasPrefix + net.Name},
-			Source:    api.HostDeviceSource{Address: &hostPCIAddress1},
-			Type:      "pci",
-			Managed:   "no",
-			BootOrder: &api.BootOrder{Order: *iface.BootOrder},
-		}
-		Expect(devices, err).To(Equal([]api.HostDevice{expectHostDevice1}))
+			callback := r.callback.(libvirt.DomainEventDeviceRemovedCallback)
+			callback(nil, nil, &libvirt.DomainEventDeviceRemoved{DevAlias: api.UserAliasPrefix + sriov.AliasPrefix})
+
+			Expect(done).To(BeTrue())
+			Eventually(func() error { return err }, 3*time.Second).Should(Succeed())
+		})
 	})
 })
+
+type deviceDetacherStub struct {
```

Comment:
Dont we need to use "github.com/golang/mock/gomock" like in other unit test?
[generated_mock_network.go](https://github.com/kubevirt/kubevirt/blob/master/pkg/virtlauncher/virtwrap/network/generated_mock_network.go)
[network_test.go](https://github.com/kubevirt/kubevirt/blob/master/pkg/virtlauncher/virtwrap/network/network_test.go)